docs/docs/typelevel.md

```diff
@@ -0,0 +1,449 @@
+nth(as, 1)      -->  { val a$1 = as._2; a$1._1 }
+```
+If tuples get large, so do the expansions. For instance, the size of the expansion of a valid selection `nth(xs, n)` is proportional to `n`. We will show later a scheme to avoid this code blowup using `erased` values.
+
+The following expressions would all give compile-time errors since a toplevel `match` could not be reduced:
+```scala
+concat(tp, bs)
+nth(tp, 0)
+nth(as, 2)
+nth(as -1)
+```
+It's possible to add more cases to a toplevel match, thereby moving an error from compile-time to runtime. For instance, here is a version of `nth` that throws a runtime error in case the index is out of bounds:
+```scala
+transparent def nthDynamic(xs: Tuple, n: Int): Any = xs match {
+  case (x, _)   if n == 0 => x
+  case (_, xs1) if n > 0  => nthDynamic(xs1, n - 1)
+  case () => throw new IndexOutOfBoundsException
+}
+```
+Here is an expansion of `nthDynamic` applied to a tuple `as: (Int, String)` and a negative index. For clarity we have added the computed types of the intermediate values `as$i`.
+```
+        nthDynamic(as, -1)
+    ->
+        { val as$0: (String, ()) = as._1
+          nthDynamic(as$0, -2)
+        }
+    ->
+        { val as$0: (String, ()) = as._1
+          { val as$1: () = as$0._1
+            nthDynamic(as$1, -3)
+          }
+        }
+    ->
+        throw new IndexOutOfBoundsException
+```
+So the end result of the expansion is the expression `throw new IndexOutOfBoundsException`, which has type `Nothing`. It is important to note that programs are treated as _data_ in this process, so the exception will not be thrown at compile time, but only if the program is run after it compiles without errors.
+
+**Rewrite Rules** The following rewrite rules are performed when simplifying inlined bodies:
+
+ - constant folding
+ - evaluation of pattern matches in toplevel match expressions
+ - reduction of projections
+ - evaluation of if-then-else with constant expressions
+
+(to be expanded)
+
+## Matching on Types
+
+We have seen so far transparent functions that take terms (tuples and integers) as parameters. What if we want to base case distinctions on types instead? For instance, one would like to be able to write a function `defaultValue`, that, given a type `T`
+returns optionally the default value of `T`, if it exists. In fact, we can already express
+this using ordinary match expressions and a simple helper function, `scala.typelevel.anyValue`, which is defined as follows:
+```scala
+def anyValue[T]: T = ???
+```
+The `anyValue` function _pretends_ to return a value of its type argument `T`. In fact, it will always raise a `NotImplementedError` exception instead.
+This is OK, since the function is intended to be used only on the type-level; it should never be executed at run-time.
+
+Using `anyValue`, we can then define `defaultValue` as follows:
+```scala
+transparent def defaultValue[T]: Option[T] = anyValue[T] match {
```

Comment:
It's really nice like this! The equivalent in `TypeOf` terms would use `scala.Predef.valueOf` (described in [SIP-23](https://docs.scala-lang.org/sips/42.type.html)):

```scala
transparent def defaultValue[T] = ... // unchanged

val defaultByte: Option[0L] = valueOf[{ defaultValue[Long] }]
```